### PR TITLE
LibGfx/PNG: Reject image with corrupted chunks

### PR DESCRIPTION
**Pull request by LucasChollet**
Checking the crc allows the decoder to reject corrupted images instead
of panicking while trying to decode them.

Fixes https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=58055&sort=-opened&can=1&q=proj%3Aserenity%20status%3DNew

While verifying the checksum is required by the standard and thus a
correct thing to do, this is technically a workaround for the
above-mentioned issue. The issue is caused by an infinite-loop in
`ZlibDecompressor::decompress_all`, and this patch allows the decoder to
return before entering that function.

### CODE REVIEW

**Comment by timschumi** [comment]
Checking the CRC32 is fine, but it's unrelated to the OSS-Fuzz issue.

If I recall correctly, the linked OSS-Fuzz issue is what I tried to fix in #18448, but that in turn currently blocks on #18940).

**Comment by LucasChollet** [comment]
> Checking the CRC32 is fine, but it's unrelated to the OSS-Fuzz issue.

That's not entirely true, it will close the OSS-Fuzz issue. This PR enhance the PNG decoder resilience by rejecting corrupted chunks, including chunks that make our zlib decompressor go crazy. While I agree that this will not solve the underlying issue, this one should be trigerred by Zlib's fuzzer (and it probably already found [it](https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=58047&sort=-opened&can=1&q=proj%3Aserenity%20status%3DNew)).

Don't hesitate to propose a better commit description if it's not clear.

**Comment by timschumi** [comment]
> Don't hesitate to propose a better commit description if it's not clear.

I'd just remove the mention of the OSS-Fuzz issue.

The patch itself is fine (at least in intent, haven't looked closely at the actual code), but the only relation that it has to the linked issue is that it essentially papers over it (by hiding behind the fuzzers general inability to generate a matching CRC32). Yes, it will get closed automatically because that particular reproducer doesn't time out anymore, but I wouldn't call it "fixed".

**Comment by Zaggy1024** [comment]
I would honestly argue against using a checksum to validate an image, at least without good reason to implement it. If anything, preventing it from decoding chunks when they don't match a checksum will just make it vastly more difficult for a fuzzer to find an issue, while it doesn't actually prevent someone from exploiting the issue by modifying the checksum. Additionally, the fuzzer will eventually find a way to reproduce this issue, probably taking less time the more sophisticated the fuzzer.

I think it would be preferable to at least attempt to decode an image regardless of whether we think it will fail to decode, since it could come out looking largely correct depending on the type of corruption. I don't know how likely that is in the case of PNG specifically, but I think in any case if the only way to prevent some issue from occurring is to reject an image based on a checksum, the format itself is flawed.

**Comment by LucasChollet** [comment]
> I think it would be preferable to at least attempt to decode an image regardless of whether we think it will fail to decode, since it could come out looking largely correct depending on the type of corruption.

I'm really not sure about that, knowing that something is wrong but still serving an image like nothing happened is really goofy IMO.

File: Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp [hunk 1]
```diff
@@ -1209,49 +1210,52 @@ static bool process_chunk(Streamer& streamer, PNGLoadingContext& context)
...
     }
     dbgln_if(PNG_DEBUG, "Chunk type: '{}', size: {}, crc: {:x}", chunk_type, chunk_size, chunk_crc);
 
-    if (!strcmp((char const*)chunk_type, "IHDR"))
+    if (auto const computed_crc = Crypto::Checksum::CRC32(chunk_type_and_data).digest(); computed_crc != chunk_crc) {
+        dbgln_if(PNG_DEBUG, "Checksum failed expected crc: {:x}, got: {:x}", chunk_crc, computed_crc);
+        return false;
+    }
+
+    if (chunk_type == "IHDR"sv)
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp:1234]: this down here should be a separate commit (and maybe separate PR). In fact, maybe you want to revive #17519? :)

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/PNGLoader.cpp:1234]: I do it if you give your opinion on this PR :)

**Comment by Zaggy1024** [comment]
> I'm really not sure about that, knowing that something is wrong but still serving an image like nothing happened is really goofy IMO.

Perhaps I'm in the minority given that I'm focused on video, where errors ideally would be able to be ignored in all cases so that it can continue playback. My argument though would be that at least in the case of displaying a corrupted image from a local file, the only option for a user would be to try to recover whatever is still readable. For loading images from the web, it makes more sense to throw an error since the error could be caused by an issue in transferring the data, but I would doubt that the checksum is the only way to determine when this is the case.

Out of curiosity, I just did a quick check of the behavior of Firefox vs Chromium with an off-by-one in random locations in a file, and it looks like they differ on this. Chromium will decode as much of the image as it can up until and after the error, outputting an image, and Firefox only retrieves the intrinsic size of the image and then reports that it is corrupted inside the img element.

**Comment by GMTA** [comment]
Some notes / thoughts:

* Verifying PNG data with an embedded checksum should probably be optional so we're not hindering the fuzzer. That would also mean that it doesn't resolve the mentioned issue.
* For the end user, being able to display an image even if some bits got corrupted can be very useful.
* For images coming from a 3rd party (i.e. the web), we probably want to stop decoding as soon as something's wrong.

**Pull request closed without merging**